src/overlays/actors/ovl_En_Bom_Chu/z_en_bom_chu.c

```diff
@@ -32,50 +37,539 @@ const ActorInit En_Bom_Chu_InitVars = {
+        }
+    }
+
+    this->actor.draw = NULL;
+    this->actionFunc = EnBomChu_WaitForDeath;
+}
+
+void EnBomChu_WaitForDeath(EnBomChu* this, GlobalContext* globalCtx) {
+    if (this->timer == 0) {
+        Actor_MarkForDeath(&this->actor);
+    }
+}
+
+/**
+ * Transform coordinates from model space to world space, according to current orientation.
+ * `posModel` is expected to already be at world scale (1/100 compared to model scale)
+ */
+void EnBomChu_ModelToWorld(EnBomChu* this, Vec3f* posModel, Vec3f* dest) {
+    f32 x = posModel->x + this->visualJitter;
+
+    dest->x = this->actor.world.pos.x + (this->axisLeft.x * x) + (this->axisUp.x * posModel->y) +
+              (this->axisForwards.x * posModel->z);
+    dest->y = this->actor.world.pos.y + (this->axisLeft.y * x) + (this->axisUp.y * posModel->y) +
+              (this->axisForwards.y * posModel->z);
+    dest->z = this->actor.world.pos.z + (this->axisLeft.z * x) + (this->axisUp.z * posModel->y) +
+              (this->axisForwards.z * posModel->z);
+}
+
+void EnBomChu_SpawnRipplesAndSplashes(EnBomChu* this, GlobalContext* globalCtx, f32 y, s32 spawnExtraRipples) {
+    s32 pad;
+    Vec3f pos;
+
+    pos.x = this->actor.world.pos.x;
+    pos.y = y;
+    pos.z = this->actor.world.pos.z;
+
+    EffectSsGRipple_Spawn(globalCtx, &pos, 70, 500, 0);
+
+    if (spawnExtraRipples) {
+        EffectSsGRipple_Spawn(globalCtx, &pos, 70, 500, 4);
+        EffectSsGRipple_Spawn(globalCtx, &pos, 70, 500, 8);
+    } else {
+        pos.x -= this->axisForwards.x * 10.0f;
+        pos.z -= this->axisForwards.z * 10.0f;
+    }
+
+    pos.y += 5.0f;
+    EffectSsGSplash_Spawn(globalCtx, &pos, NULL, NULL, 1, 450);
+}
+
+void EnBomChu_HandleNonSceneCollision(EnBomChu* this, GlobalContext* globalCtx) {
+    Vec3f originalWorldPos;
+    Vec3f posA;
+    Vec3f posB;
+    Vec3f originalAxisUp;
+    s16 yaw;
+    f32 sin;
+    f32 cos;
+    f32 tempX;
+    CollisionPoly* poly = NULL;
+    s32 bgId = BGCHECK_SCENE;
+    s32 isFloorPolyValid;
+
+    Math_Vec3f_Copy(&originalWorldPos, &this->actor.world.pos);
+    Math_Vec3f_Copy(&originalAxisUp, &this->axisUp);
+    yaw = this->actor.shape.rot.y;
+    BgCheck2_UpdateActorAttachedToMesh(&globalCtx->colCtx, this->actor.floorBgId, &this->actor);
+
+    if (yaw != this->actor.shape.rot.y) {
+        yaw = this->actor.shape.rot.y - yaw;
+
+        sin = Math_SinS(yaw);
+        cos = Math_CosS(yaw);
+
+        tempX = this->axisForwards.x;
+        this->axisForwards.x = (this->axisForwards.z * sin) + (cos * tempX);
+        this->axisForwards.z = (this->axisForwards.z * cos) - (sin * tempX);
+
+        tempX = this->axisUp.x;
+        this->axisUp.x = (this->axisUp.z * sin) + (cos * tempX);
+        this->axisUp.z = (this->axisUp.z * cos) - (sin * tempX);
+
+        tempX = this->axisLeft.x;
+        this->axisLeft.x = (this->axisLeft.z * sin) + (cos * tempX);
+        this->axisLeft.z = (this->axisLeft.z * cos) - (sin * tempX);
```

Comment:
Are these multiplications really all reversed in the different addends/subtrahends?

Reply:
Ugh, I hate it when that happens, so messy 😩

Reply:
Disregard what I said before, you *can* make it match with them in the same order, but that order *has to be* sin/cos first.